path_dynamic_creds.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate username: %w", err)
+	}
+	password, err := b.GeneratePassword(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the template
+	now := time.Now()
+	exp := now.Add(dRole.DefaultTTL)
+	templateData := dynamicTemplateData{
+		Username:              username,
+		Password:              password,
+		DisplayName:           req.DisplayName,
+		RoleName:              roleName,
+		IssueTime:             now.Format(time.RFC3339),
+		IssueTimeSeconds:      now.Unix(),
+		ExpirationTime:        exp.Format(time.RFC3339),
+		ExpirationTimeSeconds: exp.Unix(),
+	}
+	createLDIF, err := applyTemplate(dRole.CreationLDIF, templateData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply creation_ldif template: %w", err)
+	}
+
+	// Parse the raw LDIF & run it against the LDAP client
+	entries, err := ldif.Parse(createLDIF)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse generated LDIF: %w", err)
+	}
+
+	addReq := getAddRequest(entries.Entries[0])
+
+	err = b.client.Add(config.LDAP, addReq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create LDAP entry: %w", err)
+	}
+	respData := map[string]interface{}{
+		"username": username,
+		"password": password,
+		"dn":       addReq.DN,
+	}
+	internal := map[string]interface{}{
+		"name":     roleName,
+		"username": username,
+		"password": password,
+		"dn":       addReq.DN,
+	}
+	resp := b.Secret(secretCredsType).Response(respData, internal)
+	resp.Secret.TTL = dRole.DefaultTTL
+	resp.Secret.MaxTTL = dRole.MaxTTL
+
+	return resp, nil
+}
+
+func getAddRequest(req *ldif.Entry) *ldap.AddRequest {
+	if req.Add != nil {
+		return req.Add
+	}
+
+	if req.Entry == nil {
+		return nil
+	}
+
+	// Attempt to convert the Entry to an AddRequest
+	attributes := make([]ldap.Attribute, 0, len(req.Entry.Attributes))
+	for _, entryAttribute := range req.Entry.Attributes {
+		attribute := ldap.Attribute{
+			Type: entryAttribute.Name,
+			Vals: entryAttribute.Values,
+		}
+		attributes = append(attributes, attribute)
+	}
+	addReq := &ldap.AddRequest{
+		DN:         req.Entry.DN,
+		Attributes: attributes,
+		Controls:   nil,
+	}
+	return addReq
+}
+
+func (b *backend) secretCredsRenew() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		// Retrieve the role to ensure it still exists. If it doesn't, this will reject the renewal request.
+		roleNameRaw, ok := req.Secret.InternalData["name"]
+		if !ok {
+			return nil, fmt.Errorf("missing role name")
+		}
+
+		roleName := roleNameRaw.(string)
+		dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve dynamic role: %w", err)
+		}
+		if dRole == nil {
+			return nil, fmt.Errorf("unable to renew: role does not exist")
+		}
+
+		resp := &logical.Response{
+			Secret: req.Secret,
+		}
+		return resp, nil
```

Comment:
Don't we have to extend/update the `resp.Secret.TTL` (and `resp.Secret.MaxTTL`) value based on the info we got from `dRole`?

Reply:
I don't think we do. I have little experience with lease TTLs, but from my testing this doesn't appear to be needed:

```
$ vault write openldap/role/testrole creation_ldif=... deletion_ldif=... default_ttl=1m max_ttl=5m


$ # Credentials command removed here for brevity since I have a bunch of inline stuff going on with it, but the output is:
Lease ID: openldap/cred/testrole/82qMbGXOHKlCb2o3cRXN6els
DN: cn=v_token_testrole_HoL1VANHfN_1611097161,ou=users,dc=learn,dc=example
Password: JMKxJBm4yPUHjxeE7zGfxV0DnfPBnnjwudNRNdUNBqtWRcwb7H5Pj5xr75Ty0FZX
User found


$ curl -s --header "X-Vault-Token: $(vault token lookup -format=json | jq -r .data.id)" -X PUT --data "{\"lease_id\":\"${LEASE_ID}\"}" $VAULT_ADDR/v1/sys/leases/lookup | jq
{
  <snip>
  "data": {
    "expire_time": "2021-01-19T16:00:21.812546-07:00",
    "id": "openldap/cred/testrole/82qMbGXOHKlCb2o3cRXN6els",
    "issue_time": "2021-01-19T15:59:21.812545-07:00",
    "last_renewal": null,
    "renewable": true,
    "ttl": 53
  },
  <snip>
}


$ $ vault lease renew -increment=5m "${LEASE_ID}"
WARNING! The following warnings were returned from Vault:

  * TTL of "5m" exceeded the effective max_ttl of "4m31s"; TTL value is capped
  accordingly

Key                Value
---                -----
lease_id           openldap/cred/testrole/82qMbGXOHKlCb2o3cRXN6els
lease_duration     4m31s
lease_renewable    true


$ curl -s --header "X-Vault-Token: $(vault token lookup -format=json | jq -r .data.id)" -X PUT --data "{\"lease_id\":\"${LEASE_ID}\"}" $VAULT_ADDR/v1/sys/leases/lookup | jq
{
  <snip>
  "data": {
    "expire_time": "2021-01-19T16:04:21.189962-07:00",
    "id": "openldap/cred/testrole/82qMbGXOHKlCb2o3cRXN6els",
    "issue_time": "2021-01-19T15:59:21.812545-07:00",
    "last_renewal": "2021-01-19T15:59:50.189962-07:00",
    "renewable": true,
    "ttl": 266
  },
  <snip>
}
```

Reply:
I believe that the update on the `TTL` and `MaxTTL` values is to pass the latest values from the role entry to the renewal logic if they got updated. So if you were to update the role's `ttl` (or `max_ttl`) to say 1m right before that renewal call, it would cap it accordingly.

Reply:
Oh I see what you're saying - if the role itself is modified while leases are active. I'll take a look into this.

Reply:
Done. This looks like it works as it should now. The only thing to note is that if I generate creds from a role, then update the role to have a `max_ttl` less than the current elapsed time the existing credentials aren't immediately revoked. They continue to exist until the original expiration. Is this our standard behavior? This would require operators to revoke existing leases (or wait for the original TTL). Or should there be an extra step somewhere to auto-revoke?

Reply:
That's an interesting case, and it's one where I'm not sure whether the correct thing should be to immediately revoke or simply let it cap to its current remaining TTL. I think the behavior that exists today is on the latter, which is what you're observing.